sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -159,6 +262,97 @@ string_class ProgramManager::getProgramBuildLog(const RT::PiProgram &Program) {
   return Log;
 }
 
+struct ImageDeleter {
+  void operator()(DeviceImage *I) {
+    delete[] I->BinaryStart;
+    delete I;
+  }
+};
+
+ProgramManager::ProgramManager() {
+  const char *SpvFileC = std::getenv(UseSpvEnv.c_str());
+  // If a SPIRV file is specified with an environment variable,
+  // register the corresponding image
+  if (SpvFileC)
+    SpvFile = SpvFileC;
+  if (!SpvFile.empty()) {
+    // The env var requests that the program is loaded from a SPIRV file on disk
+    std::ifstream File(SpvFile, std::ios::binary);
+
+    if (!File.is_open())
+      throw runtime_error(std::string("Can't open file specified via ") +
+                          UseSpvEnv + ": " + SpvFile);
+    File.seekg(0, std::ios::end);
+    size_t Size = File.tellg();
+    auto *Data = new unsigned char[Size];
+    File.seekg(0);
+    File.read(reinterpret_cast<char *>(Data), Size);
+    File.close();
+
+    if (!File.good()) {
+      delete[] Data;
+      throw runtime_error(std::string("read from ") + SpvFile +
+                          std::string(" failed"));
+    }
+    std::unique_ptr<DeviceImage, ImageDeleter> ImgPtr(new DeviceImage(),
+                                                      ImageDeleter());
+    ImgPtr->Version = PI_DEVICE_BINARY_VERSION;
+    ImgPtr->Kind = PI_DEVICE_BINARY_OFFLOAD_KIND_SYCL;
+    ImgPtr->DeviceTargetSpec = PI_DEVICE_BINARY_TARGET_UNKNOWN;
+    ImgPtr->BuildOptions = "";
+    ImgPtr->ManifestStart = nullptr;
+    ImgPtr->ManifestEnd = nullptr;
+    ImgPtr->BinaryStart = Data;
+    ImgPtr->BinaryEnd = Data + Size;
+    ImgPtr->EntriesBegin = nullptr;
+    ImgPtr->EntriesEnd = nullptr;
+    // TODO the environment variable name implies that the only binary format
+    // it accepts is SPIRV but that is not the case, should be aligned?
+    ImgPtr->Format = getFormat(*ImgPtr);
+
+    m_DeviceImages[OSUtil::ExeModuleHandle][SpvFileKSId].reset(
+        new std::vector<DeviceImage *>({ImgPtr.get()}));
+
+    m_OrphanDeviceImages.emplace_back(std::move(ImgPtr));
+
```

Comment:
Shouldn't we do this operations under a mutex?

Reply:
There's no need to, since all access to the private members of `ProgramManager` is guaranteed to be blocked until its construction is finished.

Reply:
That's right. Sure, I'll leave a comment here.